### ISSUE TITLE
[sklearn#34563] asv benchmarks are broken with latest asv_runner

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#316]
**Mirrored from [scikit-learn#34563](https://github.com/scikit-learn/scikit-learn/issues/34563)**

Original title: asv benchmarks are broken with latest asv_runner
Original labels: Needs Triage
Created: 2026-07-24T19:20:47Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

asv_runner has a fun hack now:

```python
    @staticmethod
    def _is_parameter_free_setup(setup):
        """
        True if ``setup`` can be called with no positional arguments.
        Used so module-level ``setup(*args, **kwargs)`` runs before
        ``setup_cache`` while class ``setup(self, n)`` waits for ``do_setup``.
        """
        try:
            inspect.signature(setup).bind()
        except TypeError:
            return False
        return True
```

Unfortunately this detects the `setup()` function in `asv_benchmarks/benchmarks/common.py` as not needing parameters, which breaks benchmarks. The symptom is benchmarks failing to run with error about `setup_cache failed`.

A hacky workaround:

```diff
diff --git a/asv_benchmarks/benchmarks/common.py b/asv_benchmarks/benchmarks/common.py
index c12da551010..eff964a955e 100644
--- a/asv_benchmarks/benchmarks/common.py
+++ b/asv_benchmarks/benchmarks/common.py
@@ -159,8 +159,12 @@ class Estimator(ABC):
             with est_path.open(mode="wb") as f:
                 pickle.dump(estimator, f)
 
-    def setup(self, *params):
+    def setup(self, param1, *params):
         """Generate dataset and load the fitted estimator"""
+        # Stupid hack to workaround fact asv_runner uses ability to call
+        # without any params as semantically significant:
+        params = (param1,) + params
+
         # This is run once per combination of parameters and per repeat so we
         # need to avoid doing expensive operations there.
```

Perhaps this should be fixed upstream, but recording this here in case someone else needs to run benchmarks.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/107 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
